File: runtime/lua/smelt/plugins/inspect.lua
-- Optional plugin: `/inspect` opens a local web UI for browsing sessions,
-- their history, and provider request/response audit data.

local M = {}
local notify = smelt.notify.scoped("inspect")

-- Start the inspector server off the main thread and return its URL, or
-- return nil plus an error string on failure.
function smelt.inspect.start()
	local result = smelt.task.external(function(id)
		smelt.inspect.__start(id)
	end)
	if result.ok then
		return result.url
	end
	return nil, result.error
end

-- Stop the running inspector server.
function smelt.inspect.stop()
	local result = smelt.task.external(function(id)
		smelt.inspect.__stop(id)
	end)
	return result.ok, result.error
end

-- Open the inspector URL in a browser when the host environment can do so.
function smelt.inspect.open(url)
	return smelt.inspect.__open_url(url)
end

function M.setup()
	smelt.cmd.register("inspect", function(_arg)
		local url, err = smelt.inspect.url()
		if not url then
			url, err = smelt.inspect.start()
		end
		if url then
			local opened = smelt.inspect.open(url)
			if opened.opened then
				notify.info("Opened UI: " .. url)
			elseif opened.error then
				notify.error("UI: " .. url .. " (could not open browser: " .. opened.error .. ")")
			elseif opened.reason then
				notify.info("UI: " .. url .. " (browser auto-open unavailable: " .. opened.reason .. ")")
			else
				notify.info("UI: " .. url)
			end
		else
			notify.error(err or "failed to start server")
		end
	end, {
		desc = "Open the session inspector web UI in a browser.",
		args = {},
		while_busy = true,
	})
end

M.setup()
return M

File: crates/tui/src/lua/api/inspect.rs
//! Lua bindings for the `/inspect` session introspection web UI.

use mlua::prelude::*;
use smelt_core::lua::doc::Tier;
use smelt_core::lua::module::LuaMod;

pub(super) fn register(
    lua: &Lua,
    smelt: &mlua::Table,
    shared: &std::sync::Arc<crate::lua::LuaShared>,
) -> LuaResult<()> {
    let inspect = LuaMod::advanced(
        lua,
        smelt,
        "inspect",
        "Local session introspection web UI. `smelt.inspect.url()` reads the running server URL; the optional `inspect` plugin adds `smelt.inspect.start()`, `smelt.inspect.stop()`, and `smelt.inspect.open()`.",
        Tier::UiHost,
    )?;

    inspect.private_live_only_fn(
        "__start",
        &["task_id"],
        move |_, task_id: u64| -> LuaResult<()> {
            crate::lua::with_platform_host(|host| host.start_inspect_server(task_id));
            Ok(())
        },
    )?;

    inspect.private_live_only_fn(
        "__stop",
        &["task_id"],
        move |_, task_id: u64| -> LuaResult<()> {
            crate::lua::with_platform_host(|host| host.stop_inspect_server(task_id));
            Ok(())
        },
    )?;

    let open_context = std::sync::Arc::clone(&shared.core);
    inspect.private_live_only_fn(
        "__open_url",
        &["url"],
        move |lua, url: String| -> LuaResult<mlua::Table> {
            let result = lua.create_table()?;
            let cwd = open_context.evaluation_cwd();
            match engine::opener::open_url_if_available_in(&url, &cwd) {
                engine::opener::OpenResult::Opened => {
                    result.set("ok", true)?;
                    result.set("opened", true)?;
                }
                engine::opener::OpenResult::Unavailable(reason) => {
                    result.set("ok", false)?;
                    result.set("opened", false)?;
                    result.set("reason", reason)?;
                }
                engine::opener::OpenResult::Failed(err) => {
                    result.set("ok", false)?;
                    result.set("opened", false)?;
                    result.set("error", err)?;
                }
            }
            Ok(result)
        },
    )?;

    inspect.fn_(
        "url",
        "Return the URL of the running inspector server, or nil if it is not running.",
        &[],
        move |lua, ()| -> LuaResult<mlua::Value> {
            crate::lua::try_with_platform_host(|host| host.inspect_server_url().into_lua(lua))
                .unwrap_or(Ok(mlua::Value::Nil))
        },
    )?;

    Ok(())
}
